include/boost/json/impl/monotonic_resource.ipp

```diff
@@ -79,35 +97,69 @@ monotonic_resource::
 }
 
 monotonic_resource::
-monotonic_resource() noexcept
-    : block_size_(64 * 1024)
+monotonic_resource(
+    std::size_t max_alloc,
+    std::size_t initial_size) noexcept
+      : max_alloc_(max_alloc),
+        block_size_(initial_size)
 {
 }
 
 monotonic_resource::
 monotonic_resource(
-    std::size_t block_size) noexcept
-    : block_size_(block_size)
+  void* buffer,
+  std::size_t buffer_size,
+  std::size_t max_alloc) noexcept
+    : initial_(buffer, buffer_size),
+      max_alloc_(max_alloc),
+      block_size_(1024)
 {
 }
 
+void
+monotonic_resource::
+clear() noexcept
+{
+    // reset the initial block (if any)
+    if (initial_.base)
+      initial_.top = initial_.base_ptr();
+    // reset all the blocks managed
+    // by this object (if any)
+    for (auto b = head_; b;)
+    {
+      auto next = b->next;
+      b->top = b->base_ptr();
+      b = next;
+    }
+    allocated_ = 0;
+}
+
 void*
 monotonic_resource::
 do_allocate(
     std::size_t n,
     std::size_t align)
 {
+    // use the initial block if possible
+    if(initial_.base)
+        if (auto p = alloc_in_block(initial_, n, align))
+            return p;
+    // cap the number of dynamically allocated bytes
+    if(allocated_ + n > max_alloc_)
+        throw std::bad_alloc();
```

Comment:
This needs to be like `detail::raise_bad_alloc` and then in that function use `BOOST_THROW_EXCEPTION`, see:
https://github.com/CPPAlliance/json/blob/4a6a6e7cc3f02f6980e68557b9843095a578934c/include/boost/json/impl/except.ipp#L42